modules/programs/fish.nix

```diff
@@ -0,0 +1,155 @@
+    # TODO: environment.loginShellInit
+  '';
+
+  interactiveForeignEnv = pkgs.writeText "fish-interactive-foreign-env" ''
+    ${cfge.interactiveShellInit}
+  '';
+
+  shell = pkgs.runCommand pkgs.fish.name
+    { buildInputs = [ pkgs.makeWrapper ]; }
+    ''
+      source $stdenv/setup
+
+      mkdir -p $out/bin
+      makeWrapper ${pkgs.fish}/bin/fish $out/bin/fish
+    '';
+
+  fishAliases = concatStringsSep "\n" (
+    mapAttrsFlatten (k: v: "alias ${k} '${v}'") cfg.shellAliases
+  );
+
+  fishVariables =
+    mapAttrsToList (n: v: ''set -x ${n} "${v}"'') cfg.variables;
+
+in
+
+{
+
+  options = {
+
+    programs.fish = {
+
+      enable = mkOption {
+        type = types.bool;
+        default = false;
+        description = ''
+          Whether to configure fish as an interactive shell.
+        '';
+      };
+
+      variables = mkOption {
+        type = types.attrsOf (types.either types.str (types.listOf types.str));
+        default = {};
+        description = ''
+          A set of environment variables used in the global environment.
+          These variables will be set on shell initialisation.
+          The value of each variable can be either a string or a list of
+          strings.  The latter is concatenated, interspersed with colon
+          characters.
+        '';
+        apply = mapAttrs (n: v: if isList v then concatStringsSep ":" v else v);
+      };
+
+      shellAliases = mkOption {
+        type = types.attrs;
+        default = cfge.shellAliases;
+        description = ''
+          Set of aliases for fish shell. See <option>environment.shellAliases</option>
+          for an option format description.
+        '';
+      };
+
+      shellInit = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Shell Script code called during fish shell initialisation.
+        '';
+      };
+
+      loginShellInit = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Shell script code called during fish login shell initialisation.
+        '';
+      };
+
+      interactiveShellInit = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Shell script code called during interactive fish shell initialisation.
+        '';
+      };
+
+      promptInit = mkOption {
+        type = types.lines;
+        default = "";
+        description = ''
+          Shell script code used to initialise fish prompt.
+        '';
+      };
+
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    environment.etc."fish/config.fish".text = ''
+      # /etc/fish/config.fish: DO NOT EDIT -- this file has been generated automatically.
+
+      set fish_function_path $fish_function_path ${pkgs.fish-foreign-env}/share/fish-foreign-env/functions
+
+      set PATH ${replaceStrings [":"] [" "] config.environment.systemPath} $PATH
+
+      ${config.system.build.setEnvironment}
```

Comment:
This works because:

```fish
$ functions export
function export --description 'Set global variable. Alias for set -gx, made for bash compatibility'
	if test -z "$argv"
        set
        return 0
    end
    for arg in $argv
        set -l v (string split -m 1 "=" -- $arg)
        switch (count $v)
            case 1
                set -gx $v $$v
            case 2
                set -gx $v[1] $v[2]
        end
    end
end
```